`test_approacher.py` sometimes fails

`test_approacher.py` fails the first time of running pytest for some reason. 

However, the second etc. run of the test is succesful.

Output:

```
============================================================================================= FAILURES ==============================================================================================
__________________________________________________________________________________________ test_approacher __________________________________________________________________________________________

plt = <pytest_plt.plugin.Mock object at 0x000001D0E7C3BB08>

    def test_approacher(plt):
        def _get_signal(shift):
            return get_signal(
                parameters.sweep_amplitude.value, parameters.sweep_center.value, shift
            )

        for ref_shift in (-0.4, -0.2, 0.3):
            for target_shift in (-0.3, 0.6):
                print(f"----- ref_shift={ref_shift}, target_shift={target_shift} -----") 
                parameters = Parameters()
                control = FakeControl(parameters)

                # approaching a line at the center is too easy
                # we generate a reference signal that is shifted in some direction       
                # and then simulate that the user wants to approach a line that is not at
                # the center (this is done using get_lock_point)
                reference_signal = _get_signal(ref_shift)

                (
                    central_y,
                    target_slope_rising,
                    _,
                    rolled_reference_signal,
                    line_width,
                    peak_idxs,
                ) = get_lock_point(reference_signal, 0, len(reference_signal))

                plt.plot(reference_signal)
                plt.plot(rolled_reference_signal)

                assert abs(central_y - Y_SHIFT) < 1

                approacher = Approacher(
                    control,
                    parameters,
                    rolled_reference_signal,
                    100,
                    central_y,
                    wait_time_between_current_corrections=0,
                )

                found = False

                for i in range(100):
                    shift = target_shift * (1 + (0.025 * np.random.randn()))
                    error_signal = _get_signal(shift)[:]
                    approacher.approach_line(error_signal)

                    if parameters.sweep_amplitude.value <= 0.2:
                        found = True
                        break

>               assert found
E               assert False

tests\test_approacher.py:116: AssertionError
--------------------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------------------- 
----- ref_shift=-0.4, target_shift=-0.3 -----
write: center=0.29833984375 amp=1
write: center=0.29833984375 amp=0.5
write: center=0.296630859375 amp=0.5
write: center=0.296630859375 amp=0.25
write: center=0.2991943359375 amp=0.25
write: center=0.2991943359375 amp=0.125
found!
----- ref_shift=-0.4, target_shift=0.6 -----
write: center=-0.6328125 amp=1
write: center=-0.6328125 amp=0.5
write: center=-0.626708984375 amp=0.5
write: center=-0.626708984375 amp=0.25
write: center=-0.6209716796875 amp=0.25
write: center=-0.6209716796875 amp=0.125
found!
----- ref_shift=-0.2, target_shift=-0.3 -----
write: center=0.2900390625 amp=1
write: center=0.2900390625 amp=0.5
write: center=0.2958984375 amp=0.5
write: center=0.2958984375 amp=0.25
write: center=0.2987060546875 amp=0.25
write: center=0.2987060546875 amp=0.125
found!
----- ref_shift=-0.2, target_shift=0.6 -----
write: center=-0.60693359375 amp=1
write: center=-0.60693359375 amp=0.5
write: center=-0.60791015625 amp=0.5
========================================================================================= warnings summary ========================================================================================== 
..\..\..\miniconda3\envs\linien_dev\lib\site-packages\scipy\fft\__init__.py:97
  C:\Users\gain\miniconda3\envs\linien_dev\lib\site-packages\scipy\fft\__init__.py:97: DeprecationWarning: The module numpy.dual is deprecated.  Instead of using dual, use the functions directly from numpy or scipy.
    from numpy.dual import register_func

..\..\..\miniconda3\envs\linien_dev\lib\site-packages\scipy\sparse\sputils.py:17: 15 warnings
  C:\Users\gain\miniconda3\envs\linien_dev\lib\site-packages\scipy\sparse\sputils.py:17: DeprecationWarning: `np.typeDict` is a deprecated alias for `np.sctypeDict`.
    supported_dtypes = [np.typeDict[x] for x in supported_dtypes]

..\..\..\miniconda3\envs\linien_dev\lib\site-packages\scipy\special\orthogonal.py:81

tests/test_pid.py::test_pid
tests/test_pid.py::test_pid
  C:\Users\gain\Documents\programming\linien\tests\test_pid.py:119: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    np.int

tests/test_pid_transfer.py::test_pid_transfer
  C:\Users\gain\Documents\programming\linien\tests\test_pid_transfer.py:39: DeprecationWarning: `np.int` is a deprecated alias for the builtin `int`. To silence this warning, use `int` by itself. Doing this will not modify any behavior and is safe. When replacing `np.int`, you may wish to use e.g. `np.int64` or `np.int32` to specify the precision. If you wish to review your current use, check 
the release note link for additional information.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    x = (scale * np.array(x)).astype(np.int)

tests/test_pid_transfer.py: 24 warnings
  C:\Users\gain\Documents\programming\linien\tests\test_pid_transfer.py:94: DeprecationWarning: `np.float` is a deprecated alias for the builtin `float`. To silence this warning, use `float` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.float64` here.
  Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
    x.astype(np.float), y.astype(np.float), label=label

tests/test_pid_transfer.py::test_pid_transfer
  C:\Users\gain\Documents\programming\linien\tests\test_pid_transfer.py:66: RuntimeWarning: invalid value encountered in true_divide
    * np.log10(np.abs(p / 4096 + 10 * i / f + d * (f / 125e6) / (2**6))),

tests/test_pid_transfer.py::test_pid_transfer
  C:\Users\gain\Documents\programming\linien\tests\test_pid_transfer.py:66: RuntimeWarning: divide by zero encountered in true_divide
```

--- a/.github/workflows/pytest.yml
+++ b/.github/workflows/pytest.yml
@@ -6,6 +6,7 @@ on:
       - gateware/**
       - linien-common/**
       - linien-server/**
+      - tests/**
 
 permissions:
   contents: read

--- a/tests/test_approacher.py
+++ b/tests/test_approacher.py
@@ -102,9 +102,9 @@ def test_approacher(plt):
             )
 
             found = False
-
-            for i in range(100):
-                shift = target_shift * (1 + (0.025 * np.random.randn()))
+            rng = np.random.default_rng(seed=0)
+            for _ in range(100):
+                shift = target_shift * (1 + (0.025 * rng.standard_normal()))
                 error_signal = _get_signal(shift)[:]
                 approacher.approach_line(error_signal)
 

--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -24,6 +24,7 @@ from linien_common.common import (
 from pytest import raises
 
 Y_SHIFT = 0
+RNG = np.random.default_rng(seed=0)
 
 
 def peak(x):
@@ -44,7 +45,7 @@ def get_signal(sweep_amplitude, center, shift):
 
 
 def add_noise(spectrum, level):
-    return spectrum + (np.random.randn(len(spectrum)) * level)
+    return spectrum + (RNG.standard_normal(len(spectrum)) * level)
 
 
 def test_determine_shift_by_correlation():

--- a/tests/test_optimizer_utils.py
+++ b/tests/test_optimizer_utils.py
@@ -62,31 +62,23 @@ def test_iq():
 
     def generate_fake_data(spectrum, phase=0):
         data = np.array([])
-
         sin = get_sin(phase=phase)
-
         for point in spectrum:
             data = np.append(data, point * sin)
-
         return data
 
     def demod(data, phase=0):
         sin = get_sin(phase=phase)
-
         block_size = len(sin)
-        N_points = round(len(data) / block_size)
-
+        n_points = round(len(data) / block_size)
         demodulated_data = []
-
-        for N in range(N_points):
-            data_slice = data[N * block_size : (N + 1) * block_size]
+        for n in range(n_points):
+            data_slice = data[n * block_size : (n + 1) * block_size]
             demodulated = np.mean(sin * data_slice)
             demodulated_data.append(demodulated)
-
         return demodulated_data
 
     final_zoom_factor = 10
-
     sweep_amplitude = 1.0
     max_val = np.pi * 5 * sweep_amplitude
     x = np.linspace(-1 * max_val, 1 * max_val, 100)
@@ -100,7 +92,7 @@ def test_iq():
     spectrum3 = spectrum_for_testing(x + RNG.random() * 3)
     data3 = generate_fake_data(spectrum3, phase=RNG.integers(0, 360))  # noqa: F841
 
-    combined = data  # + data2 + data3
+    combined = data + data2 + data3
 
     def get_slope(signal):
         return get_max_slope(signal, final_zoom_factor)

--- a/tests/test_optimizer_utils.py
+++ b/tests/test_optimizer_utils.py
@@ -16,11 +16,11 @@
 # You should have received a copy of the GNU General Public License
 # along with Linien.  If not, see <http://www.gnu.org/licenses/>.
 
-from random import randint, random
-
 import numpy as np
 from linien_server.optimization.utils import get_max_slope, optimize_phase_from_iq
 from scipy.optimize import minimize_scalar
+
+RNG = np.random.default_rng(seed=0)
 
 
 def test_get_max_slope():
@@ -95,11 +95,11 @@ def test_iq():
         spectrum = spectrum_for_testing(x) * 2
         data = generate_fake_data(spectrum, phase=30)  # phase=randint(0, 360))
 
-        spectrum2 = spectrum_for_testing(x + random() * 3)
-        data2 = generate_fake_data(spectrum2, phase=randint(0, 360))  # noqa: F841
+        spectrum2 = spectrum_for_testing(x + RNG.random() * 3)
+        data2 = generate_fake_data(spectrum2, phase=RNG.integers(0, 360))  # noqa: F841
 
-        spectrum3 = spectrum_for_testing(x + random() * 3)
-        data3 = generate_fake_data(spectrum3, phase=randint(0, 360))  # noqa: F841
+        spectrum3 = spectrum_for_testing(x + RNG.random() * 3)
+        data3 = generate_fake_data(spectrum3, phase=RNG.integers(0, 360))  # noqa: F841
 
         combined = data  # + data2 + data3
 

--- a/tests/test_optimizer_utils.py
+++ b/tests/test_optimizer_utils.py
@@ -91,36 +91,33 @@ def test_iq():
     max_val = np.pi * 5 * sweep_amplitude
     x = np.linspace(-1 * max_val, 1 * max_val, 100)
 
-    for iteration in range(1):
-        spectrum = spectrum_for_testing(x) * 2
-        data = generate_fake_data(spectrum, phase=30)  # phase=randint(0, 360))
+    spectrum = spectrum_for_testing(x) * 2
+    data = generate_fake_data(spectrum, phase=30)  # phase=randint(0, 360))
 
-        spectrum2 = spectrum_for_testing(x + RNG.random() * 3)
-        data2 = generate_fake_data(spectrum2, phase=RNG.integers(0, 360))  # noqa: F841
+    spectrum2 = spectrum_for_testing(x + RNG.random() * 3)
+    data2 = generate_fake_data(spectrum2, phase=RNG.integers(0, 360))  # noqa: F841
 
-        spectrum3 = spectrum_for_testing(x + RNG.random() * 3)
-        data3 = generate_fake_data(spectrum3, phase=RNG.integers(0, 360))  # noqa: F841
+    spectrum3 = spectrum_for_testing(x + RNG.random() * 3)
+    data3 = generate_fake_data(spectrum3, phase=RNG.integers(0, 360))  # noqa: F841
 
-        combined = data  # + data2 + data3
+    combined = data  # + data2 + data3
 
-        def get_slope(signal):
-            return get_max_slope(signal, final_zoom_factor)
+    def get_slope(signal):
+        return get_max_slope(signal, final_zoom_factor)
 
-        min_result = minimize_scalar(
-            lambda phase: -1 * get_slope(demod(combined, phase=phase)),
-            method="Bounded",
-            bounds=(0, 360),
-        )
+    min_result = minimize_scalar(
+        lambda phase: -1 * get_slope(demod(combined, phase=phase)),
+        method="Bounded",
+        bounds=(0, 360),
+    )
 
-        i = demod(combined)
-        q = demod(combined, phase=90)
+    i = demod(combined)
+    q = demod(combined, phase=90)
 
-        optimized_phase, optimized_slope = optimize_phase_from_iq(
-            i, q, final_zoom_factor
-        )
+    optimized_phase, optimized_slope = optimize_phase_from_iq(i, q, final_zoom_factor)
 
-        assert abs(min_result.x - optimized_phase) <= 0.1
-        assert abs(abs(min_result.fun) - optimized_slope) / optimized_slope <= 0.001
+    assert abs(min_result.x - optimized_phase) <= 0.1
+    assert abs(abs(min_result.fun) - optimized_slope) / optimized_slope <= 0.001
 
 
 if __name__ == "__main__":

--- a/tests/test_pid_transfer.py
+++ b/tests/test_pid_transfer.py
@@ -30,11 +30,11 @@ VCD_DIR = Path(__file__).parent / "vcd"
 @pytest.mark.slow
 def test_pid_transfer(plt):
     def pid_testbench(pid):
-        np.random.seed(299792458)
+        rng = np.random.default_rng(seed=299792458)
         amplitude = 0.01
         samples = 1 << 12
 
-        x = np.random.uniform(-amplitude, amplitude, samples)
+        x = rng.uniform(-amplitude, amplitude, samples)
         scale = 2 ** (len(pid.input) - 1) - 1
         x = (scale * np.array(x)).astype(int)
         y = np.array([0] * len(x))

--- a/tests/test_robust_autolock.py
+++ b/tests/test_robust_autolock.py
@@ -39,6 +39,8 @@ from gateware.logic.autolock_utils import DynamicDelay, SumDiffCalculator
 VCD_DIR = Path(__file__).parent / "vcd"
 FPGA_DELAY_SUMDIFF_CALCULATOR = 2
 
+RNG = np.random.default_rng(seed=0)
+
 
 def peak(x):
     return np.exp(-np.abs(x)) * np.sin(x)
@@ -55,7 +57,11 @@ def atomic_spectrum(noise_level):
     smaller_peaks = (peak(x - 10) * 1024) - (peak(x + 10) * 1024)
     target_idxs = (328, 350)
     return (
-        as_int(central_peak + smaller_peaks + (np.random.randn(len(x)) * noise_level)),
+        as_int(
+            central_peak
+            + smaller_peaks
+            + (RNG.standard_normalrandn(len(x)) * noise_level)
+        ),
         target_idxs,
     )
 
@@ -66,17 +72,17 @@ def pfd_spectrum(noise_level):
     y[y > 3000] = 3000
     y[y < -3000] = -3000
     target_idxs = (220, 300)
-    return as_int(y + (np.random.randn(len(x)) * noise_level)), target_idxs
+    return as_int(y + (RNG.standard_normal(len(x)) * noise_level)), target_idxs
 
 
 def add_noise(spectrum, level):
-    return as_int(spectrum + (np.random.randn(len(spectrum)) * level))
+    return as_int(spectrum + (RNG.standard_normal(len(spectrum)) * level))
 
 
 def add_jitter(spectrum, level=None, exact_value=None):
     assert (level is not None) or (exact_value is not None)
     if exact_value is None:
-        exact_value = int(round(np.random.randn() * level))
+        exact_value = int(round(RNG.standard_normal() * level))
 
     shift = exact_value
     return np.roll(spectrum, shift)
@@ -87,8 +93,10 @@ def test_get_description(plt, debug=True):
     def get_lock_position_from_autolock_instructions_by_simulating_fpga(
         spectrum, description, time_scale, initial_spectrum, final_wait_time
     ):
-        """This function simulated the behavior of `RobustAutolock` on FPGA
-        and allows to find out whether FPGA would lock to the correct point."""
+        """
+        This function simulated the behavior of `RobustAutolock` on FPGA and allows to
+        find out whether FPGA would lock to the correct point.
+        """
         result = {}
 
         def tb(dut):
@@ -141,7 +149,8 @@ def test_get_description(plt, debug=True):
                 plt.plot(spectrum)
 
             jitters = [
-                0 if i == 0 else int(round(np.random.randn() * 50)) for i in range(10)
+                0 if i == 0 else int(round(RNG.standard_normal() * 50))
+                for i in range(10)
             ]
 
             spectra_with_jitter = [

--- a/tests/test_robust_autolock.py
+++ b/tests/test_robust_autolock.py
@@ -58,9 +58,7 @@ def atomic_spectrum(noise_level):
     target_idxs = (328, 350)
     return (
         as_int(
-            central_peak
-            + smaller_peaks
-            + (RNG.standard_normalrandn(len(x)) * noise_level)
+            central_peak + smaller_peaks + (RNG.standard_normal(len(x)) * noise_level)
         ),
         target_idxs,
     )

